File: src/main/java/cc/hyperium/internal/addons/AddonMinecraftBootstrap.java
package cc.hyperium.internal.addons;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Handles the work of loading addon main classes once we are in the loading phase where Minecraft has just become
 * accessible.
 */
public class AddonMinecraftBootstrap {
    private static final List<IAddon> LOADED_ADDONS = new ArrayList<>();

    /**
     * For each discovered addon, load the main class.
     *
     * @throws IOException If the addon's main class is invalid.
     */
    public static void init() throws IOException {
        final Map<AddonManifest, File> toLoad = AddonBootstrap.INSTANCE.getAddons();

        final List<IAddon> loaded = new ArrayList<>();

        for (Map.Entry<AddonManifest, File> addonManifestFileEntry : toLoad.entrySet()) {
            final AddonManifest addon = addonManifestFileEntry.getKey();
            try {
                Object o = Class.forName(addon.getMainClass()).newInstance();
                if (o instanceof IAddon) {
                    loaded.add((IAddon) o);
                } else {
                    throw new IOException("Main class isn't instance of IAddon");
                }
            } catch (IllegalAccessException | InstantiationException | ClassNotFoundException e) {
                e.printStackTrace();
            }
        }

        LOADED_ADDONS.addAll(loaded);
        for (IAddon addon : LOADED_ADDONS) {
            addon.onLoad();
        }
    }

    /**
     * Get all the currently loaded addons.
     *
     * @return The loaded addons main classes.
     */
    public static List<IAddon> getLoadedAddons() {
        return LOADED_ADDONS;
    }
}

File: src/main/kotlin/cc/hyperium/internal/addons/AddonMinecraftBootstrap.kt
package cc.hyperium.internal.addons

import java.io.IOException

object AddonMinecraftBootstrap {
    @JvmStatic
    val LOADED_ADDONS = ArrayList<IAddon>()
        @JvmName("getLoadedAddons") get

    @JvmStatic
    fun init() {
        try {
            if (AddonBootstrap.phase != AddonBootstrap.Phase.INIT) {
                throw IOException("Bootstrap currently at AddonBoostrap.Phase.${AddonBootstrap.phase}, it should be at INIT")
            }

            val toLoadMap = AddonBootstrap.addonManifests.map { it.name to it }.toMap().toMutableMap()

            val toLoad = toLoadMap.map { it.value }.toMutableList()

            val loaded = ArrayList<IAddon>()
            for (addon in toLoad) {
                try {
                    val o = Class.forName(addon.mainClass).newInstance()
                    if (o is IAddon) {
                        loaded.add(o)
                    } else {
                        throw IOException("Main class isn't instance of IAddon")
                    }
                } catch (e: Throwable) {
                    e.printStackTrace()
                }
            }

            LOADED_ADDONS.addAll(loaded)
            LOADED_ADDONS.forEach(IAddon::onLoad)
            AddonBootstrap.phase = AddonBootstrap.Phase.DEFAULT
        } catch (ignored: Exception) {}
    }
}
